datastore/elastic/connection.go

```diff
@@ -44,172 +44,208 @@ func init() {
 
 func (ec *elasticConnection) Put(key datastore.Key, msg datastore.JSONMessage) error {
 	logger := plog.WithFields(log.Fields{
-		"kind":  key.Kind(),
-		"id":  key.ID(),
+		"kind": key.Kind(),
+		"id":   key.ID(),
 	})
 	logger.Debug("Put")
 	traceLogger.WithField("payload", string(msg.Bytes())).Debug("Put")
 
-	//func Index(pretty bool, index string, _type string, id string, data interface{}) (api.BaseResponse, error) {
-	var raw json.RawMessage
-	raw = msg.Bytes()
-	resp, err := core.IndexWithParameters(false, ec.index, key.Kind(), key.ID(), "", msg.Version(), "", "", "", 0, "", "", false, &raw)
+	//Because of document type depreciation in ES 7.x we combine id and key
+	args := []func(*esapi.IndexRequest){
+		ec.client.Index.WithDocumentID(fmt.Sprintf("%s-%s", key.ID(), key.Kind())),
+		ec.client.Index.WithRefresh("true"),
+	}
+
+	if (msg.Version()["seqNo"] + msg.Version()["primaryTerm"]) > 0 {
+		args = append(args, ec.client.Index.WithIfPrimaryTerm(msg.Version()["primaryTerm"]),
+			ec.client.Index.WithIfSeqNo(msg.Version()["seqNo"]))
+	}
+
+	res, err := ec.client.Index(ec.index, bytes.NewReader(msg.Bytes()), args...)
+
+	defer res.Body.Close()
 	if err != nil {
 		logger.WithError(err).Error("Put failed")
-		if eserr, iseserror := err.(api.ESError); iseserror && eserr.Code == 409 {
-			// Conflict
-			return fmt.Errorf("Your changes conflict with those made by another user. Please reload and try your changes again.")
-		}
+		//TODO handle new type of ES error
+		//if eserr, iseserror := err.(api.ESError); iseserror && eserr.Code == 409 {
+		// Conflict
+		//return fmt.Errorf("Your changes conflict with those made by another user. Please reload and try your changes again.")
+		//}
 		return err
 	}
-	indices.Refresh(ec.index)
-	traceLogger.WithField("response", resp).Debug("Put")
-	if !resp.Ok {
-		return fmt.Errorf("non OK response: %v", resp)
+
+	traceLogger.WithField("response", res).Debug("Put")
+	if res.IsError() {
+		logger.Info("error while put")
+		return fmt.Errorf("non OK response: %v", res)
 	}
 	return nil
 }
 
 func (ec *elasticConnection) Get(key datastore.Key) (datastore.JSONMessage, error) {
 	logger := plog.WithFields(log.Fields{
-		"kind":  key.Kind(),
-		"id":  key.ID(),
+		"kind": key.Kind(),
+		"id":   key.ID(),
 	})
 	logger.Debug("Get")
 	traceLogger.Debug("Get")
 
-	//	func Get(pretty bool, index string, _type string, id string) (api.BaseResponse, error) {
-	//	err := core.GetSource(ec.index, key.Kind(), key.ID(), &bytes)
-	response, err := elasticGet(false, ec.index, key.Kind(), key.ID())
+	//Because of document type depreciation in ES 7.x we combine id and key
+	response, err := ec.elasticGet(fmt.Sprintf("%s-%s", key.ID(), key.Kind()))
```

Comment:
I'd like a function or method on Key that does this generation of the actual ID instead of spreading out the format everywhere.

Reply:
done